statsmodels/tsa/vecm/vecm.py

```diff
@@ -0,0 +1,557 @@
+from math import log, pi
+import numpy as np
+from numpy import hstack, vstack
+from numpy.linalg import inv, svd
+import scipy
+import scipy.stats
+from statsmodels.tools.decorators import cache_readonly
+from statsmodels.tools.tools import chain_dot
+from statsmodels.tsa.tsatools import duplication_matrix, vec, vech, unvec
+
+import statsmodels.tsa.base.tsa_model as tsbase
+
+def mat_sqrt(_2darray):
+    u_, s_, v_ = svd(_2darray, full_matrices=False)
+    s_ = np.sqrt(s_)
+    return chain_dot(u_, np.diag(s_), v_)
+
+def _endog_matrices(endog_tot, diff_lags, deterministic):
+    # p. 286:
+    p = diff_lags+1
+    y = endog_tot
+    K = y.shape[0]
+    y_1_T = y[:, p:]
+    T = y_1_T.shape[1]
+    delta_y = np.diff(y)
+    delta_y_1_T = delta_y[:, p-1:]
+    y_min1 = y[:, p-1:-1]
+    if "lt" in deterministic:
+        y_min1 = vstack((y_min1,
+                         np.arange(T)))  # p. 299
+    # p. 286:
+    delta_x = np.zeros((diff_lags*K, T))
+    for j in range(delta_x.shape[1]):
+        delta_x[:, j] = (delta_y[:, j+p-2:None if j-1<0 else j-1:-1]
+                         .T.reshape(K*(p-1)))
+    # p. 299, p. 303:
+    if "c" in deterministic:
+        delta_x = vstack((delta_x,
+                          np.ones(T)))
```

Comment:
I think it's better to move the translation from deterministic options ("nc", "lt", ..) to the two deterministic arrays to a separate function and use here just premade arrays deterministic_co and deterministic_diff.
deterministic_diff can include the seasonal component additional to any possible constant or trend.

(This way all definition for user options, whether jmulti verstion or Stata version can be isolated to a single function.)